File: jserver-security/src/main/java/com/github/gr3gdev/jserver/security/password/BCryptPasswordManager.java
package com.github.gr3gdev.jserver.security.password;

import org.bouncycastle.crypto.generators.BCrypt;

import java.nio.charset.StandardCharsets;

/**
 * BCryptPasswordManager.
 *
 * @author Gregory Tardivel
 */
public class BCryptPasswordManager {

    private final int strength;

    public BCryptPasswordManager(final int strength) {
        this.strength = strength;
    }

    /**
     * Encode password with BCrypt.
     *
     * @param password Password
     * @return String
     */
    public String encode(final String password) {
        return new String(
                BCrypt.generate(
                        password.getBytes(StandardCharsets.UTF_8),
                        new byte[16],
                        strength)
        );
    }

    /**
     * Check password.
     *
     * @param password plain text
     * @param encoded  password hashed
     * @return boolean
     */
    public boolean matches(final String password, final String encoded) {
        return encoded.equals(encode(password));
    }

}

File: jserver-security/src/main/kotlin/com/github/gr3gdev/jserver/security/password/BCryptPasswordManager.kt
package com.github.gr3gdev.jserver.security.password

import org.mindrot.jbcrypt.BCrypt
import java.security.SecureRandom

class BCryptPasswordManager(private val strength: Int) {

    private val random = SecureRandom()

    /**
     * Encode password with BCrypt.
     * @param password Password
     */
    fun encode(password: String): String {
        val salt = generateSalt()
        return BCrypt.hashpw(password, salt)
    }

    /**
     * Check password.
     * @param password plain text
     * @param encoded password hashed
     */
    fun matches(password: String, encoded: String): Boolean {
        return BCrypt.checkpw(password, encoded)
    }

    private fun generateSalt(): String {
        return BCrypt.gensalt(strength, random)
    }

}